zebra-network/src/protocol/external/addr/v1.rs

```diff
@@ -0,0 +1,169 @@
+};
+
+use crate::{
+    meta_addr::MetaAddr,
+    protocol::external::{types::PeerServices, MAX_PROTOCOL_MESSAGE_LEN},
+};
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest_derive::Arbitrary;
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use crate::protocol::external::arbitrary::addr_v1_ipv6_mapped_socket_addr_strategy;
+
+/// A version 1 node address, its advertised services, and last-seen time.
+/// This struct is serialized and deserialized into `addr` messages.
+///
+/// [Bitcoin reference](https://en.bitcoin.it/wiki/Protocol_documentation#Network_address)
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub(in super::super) struct AddrV1 {
+    /// The unverified "last seen time" gossiped by the remote peer that sent us
+    /// this address.
+    ///
+    /// See the [`MetaAddr::last_seen`] method for details.
+    untrusted_last_seen: DateTime32,
+
+    /// The unverified services for the peer at `ipv6_addr`.
+    ///
+    /// These services were advertised by the peer at `ipv6_addr`,
+    /// then gossiped via another peer.
+    ///
+    /// ## Security
+    ///
+    /// `untrusted_services` on gossiped peers may be invalid due to outdated
+    /// records, older peer versions, or buggy or malicious peers.
+    untrusted_services: PeerServices,
+
+    /// The peer's IPv6 socket address.
+    /// IPv4 addresses are serialized as an [IPv4-mapped IPv6 address].
+    ///
+    /// [IPv4-mapped IPv6 address]: https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
+    #[cfg_attr(
+        any(test, feature = "proptest-impl"),
+        proptest(strategy = "addr_v1_ipv6_mapped_socket_addr_strategy()")
+    )]
+    ipv6_addr: SocketAddrV6,
+}
+
+impl From<MetaAddr> for AddrV1 {
+    fn from(meta_addr: MetaAddr) -> Self {
+        let ipv6_addr = ipv6_mapped_socket_addr(meta_addr.addr);
+
+        let untrusted_services = meta_addr.services.expect(
+            "unexpected MetaAddr with missing peer services: \
+             MetaAddrs should be sanitized before serialization",
+        );
+        let untrusted_last_seen = meta_addr.last_seen().expect(
+            "unexpected MetaAddr with missing last seen time: \
+             MetaAddrs should be sanitized before serialization",
+        );
+
+        AddrV1 {
+            untrusted_last_seen,
+            untrusted_services,
+            ipv6_addr,
+        }
+    }
+}
+
+impl From<AddrV1> for MetaAddr {
+    fn from(addr_v1: AddrV1) -> Self {
+        MetaAddr::new_gossiped_meta_addr(
+            addr_v1.ipv6_addr.into(),
+            addr_v1.untrusted_services,
+            addr_v1.untrusted_last_seen,
+        )
+    }
+}
+
+impl ZcashSerialize for AddrV1 {
+    fn zcash_serialize<W: Write>(&self, mut writer: W) -> Result<(), std::io::Error> {
+        self.untrusted_last_seen.zcash_serialize(&mut writer)?;
+        writer.write_u64::<LittleEndian>(self.untrusted_services.bits())?;
```

Comment:
Also another idea/discussion for later: should we have an `impl Zcash{Des,S}erialize for PeerServices`?

Reply:
We'd need to implement serialization on two different wrapper types, because the `PeerServices` format is different in `addr`/`version` and `addrv2`.

Can you open a ticket?

Reply:
Done #3047.